demos/tutorials/Gibbs.py

```diff
@@ -0,0 +1,222 @@
+from cuqi.distribution import GaussianCov, Gamma, JointDistribution, GMRF, Laplace_diff
+from cuqi.sampler import Gibbs, Linear_RTO, Conjugate, UnadjustedLaplaceApproximation, ConjugateApprox
+
+np.random.seed(0)
+
+# %%
+# Forward model and data
+# ------------------------
+# We define the forward model and data.
+# Here we use a 1D deconvolution problem, so the forward model is linear,
+# that is:
+#
+# .. math::
+#    \mathbf{y} = \mathbf{A} \mathbf{x}
+#
+# where :math:`\mathbf{A}` is the convolution matrix, and :math:`\mathbf{x}` is the input signal.
+#
+# We load this example from the testproblem library of CUQIpy.
+
+# Model and data
+A, y_obs, probinfo = Deconvolution1D.get_components(phantom='square')
+
+# Get dimension of signal
+n = A.domain_dim
+
+# %%
+# Hierarchical Bayesian model
+# ---------------------------
+#
+# We define the following hierarchical model:
+#
+# .. math::
+#
+#     \begin{align}
+#         d &\sim \mathrm{Gamma}(1, 10^{-4}) \\
+#         l &\sim \mathrm{Gamma}(1, 10^{-4}) \\
+#         \mathbf{x} &\sim \mathrm{GMRF}(\mathbf{0}, d) \\
+#         \mathbf{y} &\sim \mathcal{N}(\mathbf{A} \mathbf{x}, l^{-1} \mathbf{I}_m)
+#     \end{align}
+#
+# where :math:`\mathbf{y}` is the observed data, and :math:`\mathbf{x}`
+# is the unknown signal. The hyperparameters :math:`d` and :math:`l` are
+# the precision of the prior distribution of :math:`\mathbf{x}` and
+# the noise, respectively.
+#
+# The prior distribution of :math:`\mathbf{x}` is a Gaussian Markov random
+# field (GMRF) with zero mean and precision :math:`d`. It can
+# be viewed as a Gaussian prior on the differences between neighboring
+# elements of :math:`\mathbf{x}`.
+#
+# In CUQIpy the model can be defined as follows:
+
+# Define distributions
+d = Gamma(1, 1e-4)
+l = Gamma(1, 1e-4)
+x = GMRF(np.zeros(n), lambda d: d[0])
+y = GaussianCov(A, lambda l: 1/l)
+
+# Combine into a joint distribution
+joint = JointDistribution([d, l, x, y])
+
+# Define posterior by conditioning on the data
+posterior = joint(y=y_obs)
+
+print(posterior)
+
+# %%
+# Gibbs Sampler
+# -------------
+#
+# The hierarchical model above has some important properties that we
+# can exploit to make the sampling more efficient. First, note that
+# the Gamma distribution are conjugate priors for the precision of
+# the Gaussian distributions. This means that we can efficiently sample
+# from :math:`d` and :math:`l` conditional on the other variables.
+# 
+# Second, note that the prior distribution of :math:`\mathbf{x}` is
+# a Gaussian Markov random field (GMRF) and that the distribution for
+# :math:`\mathbf{y}` is also Gaussian with a Linear operator acting
+# on :math:`\mathbf{x}` as the mean variable. This means that we can
+# efficiently sample from :math:`\mathbf{x}` conditional on the other
+# variables using the ``Linear_RTO`` sampler.
+# 
+# Taking these two facts into account, we can define a Gibbs sampler
+# that uses the ``Conjugate`` sampler for :math:`d` and :math:`l` and
```

Comment:
I think more explanation is needed of what the Conjugate sampler is, again probably better in a follow-up tutorial after a first more basic one with one or more simpler examples.

Reply:
Added #35